### PR TITLE
[#13] Implement Bahdanau attention model.

### PR DESCRIPTION
**Pull request by dansuh17**
- paper: https://arxiv.org/abs/1409.0473

### CODE REVIEW

File: image_captioning/models/attention.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        hidden_with_time_axis = tf.expand_dims(hidden_state, axis=1)
+
+        # `hidden_with_time_axis` is broadcasted over time axis (axis 1).
+        # attention_hidden_layer.shape == (batch, time, attn_hidden_units)
+        attention_hidden_layer = tf.nn.tanh(
+            self.dense1(features) + self.dense2(hidden_with_time_axis))
+
+        # Scores per time.
+        # score.shape == (batch, time, score=1)
+        score = self.value_weight(attention_hidden_layer)
```
Comment (crissed53) [inline image_captioning/models/attention.py:72]: lol, nice approach replacing dot product with a vector with a dense layer

Comment (crissed53) [reply image_captioning/models/attention.py:72]: But have you done this on purpose to have a dummy axis (score axis)? In PyTorch, I would have calculated the score with the last axis squeezed (and also the attention weights below) by directly calculating dot product with a vector with learnable weights, but I was wondering if you have left the axis on purpose

**Review by crissed53 — COMMENTED** [review]
LGTM with minor questions

File: image_captioning/models/attention.py (hunk 1 above)
Comment (dansuh17) [reply image_captioning/models/attention.py:72]: Not sure I understand your question but the preference for `Dense` over learnable vectors using [`Variable` ](https://www.tensorflow.org/api_docs/python/tf/Variable) is its convenience in weight initialization (defaulting to `glorot_uniform`) and not having to specify the input shape.

Leaving the last axis there allows me to broadcast the scores over the embedding vectors during the element-wise multiplication in `attention_weights * features` below. Not sure it works without that axis - didn't test.

**Review by crissed53 — APPROVED** [review]
LGTM
